vpr/src/route/router_lookahead_map_utils.cpp

```diff
@@ -0,0 +1,619 @@
+        Tsw += Tsw_adjust;
+        VTR_ASSERT(Tsw >= 0.f);
+        float Rsw = device_ctx.rr_switch_inf[switch_ind].R;
+        float Cnode = device_ctx.rr_nodes[set_rr_node_ind].C();
+        float Rnode = device_ctx.rr_nodes[set_rr_node_ind].R();
+
+        float T_linear = 0.f;
+        if (device_ctx.rr_switch_inf[switch_ind].buffered()) {
+            T_linear = Tsw + Rsw * Cnode + 0.5 * Rnode * Cnode;
+        } else { /* Pass transistor */
+            T_linear = Tsw + 0.5 * Rsw * Cnode;
+        }
+
+        float base_cost = 0.f;
+        if (device_ctx.rr_switch_inf[switch_ind].configurable()) {
+            base_cost = get_single_rr_cong_base_cost(set_rr_node_ind);
+        }
+
+        VTR_ASSERT(T_linear >= 0.);
+        VTR_ASSERT(base_cost >= 0.);
+        this->delay += T_linear;
+
+        this->congestion_upstream += base_cost;
+    }
+
+    /* set the cost of this node */
+    this->cost = this->delay;
+}
+
+util::PQ_Entry_Delay::PQ_Entry_Delay(
+    int set_rr_node_ind,
+    int switch_ind,
+    const util::PQ_Entry_Delay* parent) {
+    this->rr_node_ind = set_rr_node_ind;
+
+    if (parent != nullptr) {
+        auto& device_ctx = g_vpr_ctx.device();
+        float Tsw = device_ctx.rr_switch_inf[switch_ind].Tdel;
+        float Rsw = device_ctx.rr_switch_inf[switch_ind].R;
+        float Cnode = device_ctx.rr_nodes[set_rr_node_ind].C();
+        float Rnode = device_ctx.rr_nodes[set_rr_node_ind].R();
+
+        float T_linear = 0.f;
+        if (device_ctx.rr_switch_inf[switch_ind].buffered()) {
+            T_linear = Tsw + Rsw * Cnode + 0.5 * Rnode * Cnode;
+        } else { /* Pass transistor */
+            T_linear = Tsw + 0.5 * Rsw * Cnode;
+        }
+
+        VTR_ASSERT(T_linear >= 0.);
+        this->delay_cost = parent->delay_cost + T_linear;
+    } else {
+        this->delay_cost = 0.f;
+    }
+}
+
+util::PQ_Entry_Base_Cost::PQ_Entry_Base_Cost(
+    int set_rr_node_ind,
+    int switch_ind,
+    const util::PQ_Entry_Base_Cost* parent) {
+    this->rr_node_ind = set_rr_node_ind;
+
+    if (parent != nullptr) {
+        auto& device_ctx = g_vpr_ctx.device();
+        if (device_ctx.rr_switch_inf[switch_ind].configurable()) {
+            this->base_cost = parent->base_cost + get_single_rr_cong_base_cost(set_rr_node_ind);
+        } else {
+            this->base_cost = parent->base_cost;
+        }
+    } else {
+        this->base_cost = 0.f;
+    }
+}
+
+/* returns cost entry with the smallest delay */
+util::Cost_Entry util::Expansion_Cost_Entry::get_smallest_entry() const {
+    util::Cost_Entry smallest_entry;
+
+    for (auto entry : this->cost_vector) {
+        if (!smallest_entry.valid() || entry.delay < smallest_entry.delay) {
+            smallest_entry = entry;
+        }
+    }
+
+    return smallest_entry;
+}
+
+/* returns a cost entry that represents the average of all the recorded entries */
+util::Cost_Entry util::Expansion_Cost_Entry::get_average_entry() const {
```

Comment:
It might be good to write standard functions for these statistics, and then use them to make these much shorter.
There are a lot of C++ functions in `<algorithm>` that you could use, if only we were using C++20.

* min: https://en.cppreference.com/w/cpp/algorithm/min_element
* median: https://en.cppreference.com/w/cpp/algorithm/nth_element
* sum, average: https://en.cppreference.com/w/cpp/algorithm/accumulate

My strategy is to just copy these, so that we could use standard library functions if we ever move to a later standard.

Reply:
I would leave this as a future upgrade of the Expansion cost entry, as currently, we are leveraging only the smallest cost entry calculation method.